src/Symfony/Bundle/FrameworkBundle/Form/LoginType.php

```diff
@@ -0,0 +1,94 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Bundle\FrameworkBundle\Form;
+
+use Symfony\Bundle\SecurityBundle\Security\FirewallMap;
+use Symfony\Component\Form\AbstractType;
+use Symfony\Component\Form\Extension\Core\Type\PasswordType;
+use Symfony\Component\Form\Extension\Core\Type\TextType;
+use Symfony\Component\Form\FormBuilderInterface;
+use Symfony\Component\Form\FormError;
+use Symfony\Component\Form\FormEvent;
+use Symfony\Component\Form\FormEvents;
+use Symfony\Component\HttpFoundation\RequestStack;
+use Symfony\Component\OptionsResolver\OptionsResolver;
+use Symfony\Component\Routing\Generator\UrlGeneratorInterface;
+use Symfony\Component\Security\Http\Authentication\AuthenticationUtils;
+use Symfony\Component\Translation\TranslatorInterface;
+
+class LoginType extends AbstractType
+{
+    private $requestStack;
+    private $firewallMap;
+    private $authenticationUtils;
+    private $urlGenerator;
+    /**
+     * @var TranslatorInterface
+     */
+    private $translator;
+
+    public function __construct(RequestStack $requestStack, FirewallMap $firewallMap, AuthenticationUtils $authenticationUtils, UrlGeneratorInterface $urlGenerator, TranslatorInterface $translator)
+    {
+        $this->requestStack = $requestStack;
+        $this->firewallMap = $firewallMap;
+        $this->authenticationUtils = $authenticationUtils;
+        $this->urlGenerator = $urlGenerator;
+        $this->translator = $translator;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildForm(FormBuilderInterface $builder, array $options)
+    {
+        $firewallConfig = $this->firewallMap->getFirewallConfig($this->requestStack->getCurrentRequest());
```

Comment:
This looks odd to me, you can only do stuff with forms in the master request, sub-requests can't be trusted with that due to missing data.

I'm not sure how to do this properly, but the current check feels wrong to me.

Reply:
Request should be a mandatory option of `LoginType` as well as some others like the firewall name

Reply:
For the DX it would be better to keep everything in the form and IMO it's not a big deal.

I proposed to throw any kind of LogicException if the conditions to use the form are not filled.
So it would be : 
 - master request
 - security enabled
 - form_login or simple_form listener.

What do you think ?

@Pierstoval the firewall name is not needed in the options as it's the current firewall config that is provided by the FirewallMap

Reply:
Imo a default login form is not very useful if it only works for the authenticators that are least likely to be used. What if you move the checks to the authentication utils and resolve a bunch of form options in there, which can then be used to create a form type of your liking? I think it would be best if the form type was pretty "dumb" and only uses options to configure the form.

Reply:
Sounds like a good idea. I'll check that ;-)

Reply:
> * master request

I don't agree: it would be much better to keep a fully operational form even with subrequests, so manually injecting the request might be the best option to me.

> * form_login or simple_form listener.

This is unnecessary because all parameters related to `form_login` or `simple_form` are already handled by the associated security handlers.
As said in other comments, I'm not using `form_login` because I have a multidomain application connecting to legacy etc., all of this with a single login form managed by a Guard authenticator.
A built-in `LoginType` would be really interesting to me to avoid having to write my own which replicates the `form_login` one but with more flavour. If a built-in one could be security-provider-free, it'd be compatible with any authenticator, could it be `form_login`, `guard` or any other.